driven/storage/adapter_v3.go

```diff
@@ -11,51 +14,295 @@ import (
 	"go.mongodb.org/mongo-driver/mongo/options"
 )
 
+// FindGroupsV3 finds groups with filter
+func (sa *Adapter) FindGroupsV3(clientID string, filter *model.GroupsFilter) ([]model.Group, error) {
+	groupIDs := []string{}
+	var err error
+	var userID *string
+	var memberships model.MembershipCollection
+
+	groupFilter := bson.D{primitive.E{Key: "client_id", Value: clientID}}
+	findOptions := options.Find()
+
+	if filter != nil {
+		groupIDMap := map[string]bool{}
+		if len(filter.GroupIDs) > 0 {
+			for _, groupID := range filter.GroupIDs {
+				groupIDMap[groupID] = true
+			}
+		}
+
+		if filter.MemberUserID == nil && filter.MemberExternalID != nil {
+			var user model.User
+			err := sa.db.users.Find(bson.D{
+				{"client_id", clientID},
+				{"external_id", filter.MemberExternalID},
+			}, &user, nil)
+			if err != nil {
+				userID = &user.ID
+			}
+		}
+		if userID == nil && filter.MemberUserID == nil && filter.MemberID != nil {
+			membership, _ := sa.FindGroupMembershipByID(clientID, *filter.MemberID)
+			if membership != nil {
+				memberships = model.MembershipCollection{
+					Items: []model.GroupMembership{*membership},
+				}
+				userID = &membership.UserID
+				if len(groupIDMap) == 0 || groupIDMap[membership.GroupID] {
+					groupIDs = append(groupIDs, membership.GroupID)
+				}
+			}
+		}
+
+		if filter.MemberUserID != nil {
+			// find group memberships
+			memberships, err = sa.FindGroupMemberships(clientID, model.MembershipFilter{
+				UserID: filter.MemberUserID,
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			for _, membership := range memberships.Items {
+				if len(groupIDMap) == 0 || groupIDMap[membership.GroupID] {
+					groupIDs = append(groupIDs, membership.GroupID)
+				}
+			}
+		}
+
+		if filter.MemberExternalID != nil {
+			// find group memberships
+			memberships, err = sa.FindGroupMemberships(clientID, model.MembershipFilter{
+				ExternalID: filter.MemberExternalID,
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			for _, membership := range memberships.Items {
+				if len(groupIDMap) == 0 || groupIDMap[membership.GroupID] {
+					groupIDs = append(groupIDs, membership.GroupID)
+				}
+			}
+		}
+
+		if userID != nil && (filter.MemberID == nil && filter.MemberUserID == nil && filter.MemberExternalID == nil) {
```

Comment:
Quite a good point! I decided to remove it as an overcomplicated block and really hard to read. 
The idea was to cover the privacy permission, but however the intention who requests this method and for what took out of control and need to step back.

Reply:
@shurwit  @roberlander2 Let me know please your thoughts.

Reply:
Thanks @mdryankov, I think removing this block is fine. I have an additional comment on `FindGroupsV3` in general, since I believe it can be simplified to avoid some confusion and unnecessary DB requests.

I think the following can replace the 4 `if` blocks before the first `if` block modifying the filter. It looks to me that these 4 `if` blocks are used to find memberships and add their respective group IDs to `groupIDs`, so I think this can all be done with one DB request. Please let me know if I am missing something here or if you have any additional thoughts.
```
if filter.MemberUserID != nil || filter.MemberID != nil || filter.MemberExternalID != nil {
        // find group memberships
	memberships, err = sa.FindGroupMemberships(clientID, model.MembershipFilter{
		ID: filter.MemberID,
		UserID: filter.MemberUserID,
		ExternalID: filter.MemberExternalID,
	})
	if err != nil {
		return nil, err
	}

	for _, membership := range memberships.Items {
		if len(groupIDMap) == 0 || !groupIDMap[membership.GroupID] {
			groupIDs = append(groupIDs, membership.GroupID)
			groupIDMap[membership.GroupID] = true
		}
	}
}
```

Reply:
Yes. Thank you @roberlander2